Add rust/cargo build support

fabiosky opened the pull request:
Added support to build rust crates.

**Added tools**
`Tool_cargo`: This tool is the rust package manager/builder. The latest stable version of rust is downloaded and installed. In order to support both 32 and 64 bits builds, both `x86_64-pc-windows-msvc` and `i686-pc-windows-msvc` are installed.

**Added base builders**
`Rust`: Rust project builder. Handles:
* build platform (32/64 bits)
* build type (debug/release). **Note**: in rust a release build, unless specified, will not contain debug symbols. Since we do want a release with debug symbols, we request this adding `-g` to `RUSTFLAGS` enviroment variable
* commands: clean/build/test
* per-project provided cargo params `cargo_params`
* global provided params `cargo_opts` 

**Added projects**
`Dcv-color-primitives`: a rust library to perform color conversion. https://github.com/aws/dcv-color-primitives

nacho commented:
@guruDanny67 I already reviewed this locally with Fabio. Can you give it a look as well?

guruDanny67 commented:
> 
> 
> @guruDanny67 I already reviewed this locally with Fabio. Can you give it a look as well?

I take a fast look.

The first thing is that 'cargo' is missing in the 'Tool' group: we should add it so when we launch the dev-shell we have also the commands to handle rust.

The biggest problem happens when I tried to build dcv-color-primitives: in x86 or x64 mode I always got this error and the build stops :(

error: no default toolchain configured

I don't have any rust installed on my system.

Doing a couple of tests seems to me that RUSTUP_HOME & CARGO_HOME are missing (but when I put them, manually I can see only the x64 toolchain).

Putting the correct env before launching the build script, in x64 mode the color project is built but I got an error in x86 mode, in the rust compiling/setup step:

   Compiling semver v0.9.0
error[E0425]: cannot find function `_mm256_extract_epi64` in this scope
   --> src\convert_image\avx2.rs:209:28
    |
209 |     *(image as *mut i64) = _mm256_extract_epi64(z, 0);
    |                            ^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `_mm256_extract_epi16`
error: aborting due to previous error

Hope this helps

fabiosky commented:
Thanks @guruDanny67 for your feedback. I was able to reproduce the build error.

I found out and fixed the following issues:
* The gvsbuild cargo install might pollute the system. By default, a rust installation modifies the Path registry key. This is clearly not wanted.
* The builder was not using the downloaded rust/cargo. On my local machine I already have another rust installation, and the gvsbuild script erroneusly used that.
* When executing tests, the generated outputs did not have the correct name.
* dcv-color-primitives did not compile in 32-bit mode. This issue was not related to gvsbuild. I already pushed the required changes in https://github.com/aws/dcv-color-primitives/pull/3
I also added 'cargo` to groups

I reexecuted the build process on [32,64]x[debug,release] starting on a machine without any cargo installation and now all those combinations should build successfully.

Could you please have a look on that?

Thanks,
Fabio

guruDanny67 commented:
Now it works way better, great!

I can build the x86 & x64 version of the color_primitive project without errors (tested with vs2017 & vs2019).

It doesn't work for the dev-shell: the bin part is missing from the added path (e.g. 'c:\gtk-build\tools\cargo' instead of 'c:\gtk-build\tools\cargo\bin') and the environment RUSTUP_HOME & CARGO_HOME are still missing (but is not easy with the current Tool adding them, we should change the base class to have the support for it).

I think that we can merge now and fix this thing later.

Thanks

nacho commented:
Thanks for testing!
